skimage/transform/_ring_detector.pyx

```diff
@@ -0,0 +1,1184 @@
+        np.ndarray[RIJ_t ,ndim=1] votes
+        INDX_t [:,:] vote4xy
+
+    Rmin-=1
+    Rmax+=1
+    Nrows = curv.shape[0]
+    Ncols = curv.shape[1]
+    Nrads = Rmax-Rmin+1
+    directed_ridges = {}
+    counter = 0
+    votes = np.empty(Nrows*Ncols*Nrads*2, dtype=RIJ)
+
+    # iterate over all image entries (principal curv in this case).
+    # avoid dealing with the boundaries by iterating over all but the entries
+    # on the exterior.
+    for i in xrange(1,Nrows-1):         
+        for j in xrange(1,Ncols-1):
+            #
+            ###   Find Ridges:    ###
+            #
+            # treshold the curvature (note that it should be smaller than...)
+            if curv[i,j] > curv_thresh: continue
+            # perform non-minimum suppression in the least principal direction
+            cosQ,sinQ = least_principal_direction(Lrr[i,j], Lcc[i,j], Lrc[i,j])
+            if fabs(cosQ) > cos_q_pi:
+                if (curv[i,j] >= curv[i,j+1]) | (curv[i,j] >= curv[i,j-1]):
+                    continue
+            elif fabs(sinQ) > cos_q_pi:
+                if (curv[i,j] >= curv[i+1,j]) | (curv[i,j] >= curv[i-1,j]):
+                    continue
+            elif copysign(1,sinQ) == copysign(1,cosQ):
+                if (curv[i,j] >= curv[i+1,j+1]) | (curv[i,j] >= curv[i-1,j-1]):
+                    continue
+            elif copysign(1,sinQ) != copysign(1,cosQ):
+                if (curv[i,j] >= curv[i-1,j+1]) | (curv[i,j] >= curv[i+1,j-1]):
+                    continue
+           
+            # add the least principal direction to the ridge sparse array
+            directed_ridges[i,j] = cosQ, sinQ
+
+            ###   Circle Hough Transform   ###
+
+            # TODO: consider Bresenham circle (arcs) instead of Gaussian
+            # smoothing
+
+            x0 = fround(sinQ*Rmin)
+            y0 = fround(cosQ*Rmin)
+            x1 = fround(sinQ*Rmax)
+            y1 = fround(cosQ*Rmax)
+            vote4xy = vote4(Rmin, x0, y0, Rmax, x1, y1)
+
+            for sign in xrange(-1,2,2):
+            # 20130405: corrected erraneous xrange(-1,2)
+            # for the inwards and outwards without the zero (inplace)
+                for r_ in xrange(Nrads):
+                    x = <RIJ_t>(i + sign*vote4xy[r_,0])
+                    y = <RIJ_t>(j + sign*vote4xy[r_,1])
+                    r = r_+Rmin
+                    ## do not need to check for above origin due to the unsign:
+                    #if (x<0) | (x>=Nrows) | (y<0) | (y>=Ncols): break
+                    if (x>=Nrows-1) | (y>=Ncols-1): break  ## avoid outer frame error
+
+                    votes[counter] = coord2rij(r,x,y)
+                    counter+=1
+                    
+    votes = votes[:counter]
+    votes.sort()
+    Rmin+=1
+    Rmax-=1
+
+    return {'directed_ridges':directed_ridges, 'votes':votes}
+
+
+#@cython.boundscheck(True)
+#@cython.wraparound(True)
+@cython.cdivision(True)
+cpdef votes2rings(RIJ_t [:] votes,
```

Comment:
Maybe just `cdef` here and for the other helpers. Then this is completely gone from python and you don't have to worry about making sure the parameters are well defined. (because the callers already checked them).